### PR TITLE
Fixes: #243 Clicking on an uploaded image opens in new activity

### PR DESCRIPTION
**Pull request by saketkumar95**
This PR address https://github.com/zulip/zulip-android/issues/243

videotogif_2017 01 08_16 43 07

@kunall17 Review this and let me know if any changes required. :)

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@saketkumar95, it looks like you've already signed the Dropbox CLA. Thanks!

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+import com.bumptech.glide.Glide;
+import com.bumptech.glide.load.resource.drawable.GlideDrawable;
+import com.bumptech.glide.request.RequestListener;
+import com.bumptech.glide.request.target.GlideDrawableImageViewTarget;
+import com.bumptech.glide.request.target.Target;
+import com.zulip.android.R;
+
+
+/**
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: Author details are saved in the git commit author info, you can remove this one!

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    private void copyLink(String url) {
+        ClipboardManager clipboard
+                = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
+        ClipData clip = ClipData.newPlainText("link", url);
+        clipboard.setPrimaryClip(clip);
+        Toast.makeText(PhotoViewActivity.this , R.string.link_copied , Toast.LENGTH_SHORT).show();
+    }
+
+}
+
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: Too many spaces in here, please remove them!

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java (hunk 1 above)
Comment (saketkumar95) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: Can i add a comment that what actually this activity is doing?

Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: Yeah sure create a javadoc format docs!

**Comment by saketkumar95** [comment]
@kunall17 Done. Review now. :)

File: app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java [hunk 3]
```diff
@@ -381,10 +382,11 @@ public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int pos
                             .setOnClickListener(new View.OnClickListener() {
                                 @Override
                                 public void onClick(View view) {
-                                    Intent i = new Intent(Intent.ACTION_VIEW);
+                                    Intent i = new Intent(zulipApp.getApplicationContext(), PhotoViewActivity.class);
                                     i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
-                                    i.setData(Uri.parse(url));
+                                    i.putExtra(Intent.EXTRA_TEXT , url);
                                     zulipApp.startActivity(i);
+                                    ((Activity) context).overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: I think this Transition will be perfect for this purpose.
lydorea
1cjqssa
rukkfuz

https://guides.codepath.com/android/Shared-Element-Activity-Transition

Comment (saketkumar95) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: @vishwesh3 Yeah it looks good. 
@kunall17 what do you think? Can i add this?

Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: His main focus is on the title, but what's the title in our case?
I think just a pop out of the image animation (size growing) will do, if its not a big hassle?

Comment (saketkumar95) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: @kunall17 I just tried to make it a lot simple as it is on web. That's why i didn't made much good transitions. And i guess we don't need such transitions just for displaying images.  As for now i just made title as the name of pic which is fetched in url and displayed it in toolbar's title.

Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: @saketkumar95 yeah such complex not needed, thats what I said above ` if its not a big hassle` just a pop up or slide in will do!

**Comment by kunall17** [comment]
@saketkumar95 how about a download image option in the overflow I think i will be very useful?

File: app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java (hunk 3 above)
Comment (saketkumar95) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: @kunall17  `android.R.anim.fade_in` and  `android.R.anim.fade_out` is best for now i guess.  required then we can do it in future. :)

**Comment by saketkumar95** [comment]
@kunall17 I too thought of that and was about to ask. Just want to confirm this that i have to add the download feature in this commit only or in the other commit?

**Comment by kunall17** [comment]
In this PR another commit!

File: app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java (hunk 3 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java]: :+1:

**Comment by saketkumar95** [comment]
@kunall17 Review now. :)

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                    total = total + count;
+                    outputStream.write(data, 0, count);
+                    calculatedProgress = (int) total * 100 / imageLength;
+                    publishProgress(total);
+                }
+                inputStream.close();
+                outputStream.close();
+
+            } catch (MalformedURLException e) {
+                e.printStackTrace();
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: Zlog.logException(e) can be used here.

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: same comment as above :)

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java (hunk 4 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: Zlog.logException(e) can be used here.

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java (hunk 5 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: same as above.

**Comment by Sam1301** [comment]
@saketkumar95 nice work! added a comment :)

**Comment by saketkumar95** [comment]
@Sam1301 Thanks for the review. :) I have updated it now.

File: app/src/main/res/menu/photoview_options.xml [hunk 6]
```diff
@@ -11,4 +11,8 @@
         android:id="@+id/copy_link"
         android:title="Copy link" />
 
+    <item
+        android:id="@+id/download"
+        android:title="Download image" />
```
Comment (kunall17) [inline app/src/main/res/menu/photoview_options.xml]: User displayed strings in strings.xml

**Comment by kunall17** [comment]
Squash the last two commits!

**Comment by saketkumar95** [comment]
@kunall17 Done! :)

**Comment by kunall17** [comment]
Add permission for marshmallow devices

```
01-19 17:08:56.140 23489-24237/com.zulip.android.dev E/Error: oops
                                                              java.io.FileNotFoundException: /storage/emulated/0/Download/cbff0354-d47f-11e6-8bc0-2a2b3eaf014a.png: open failed: EACCES (Permission denied)
                                                                  at libcore.io.IoBridge.open(IoBridge.java:452)
                                                                  at java.io.FileOutputStream.<init>(FileOutputStream.java:87)
                                                                  at java.io.FileOutputStream.<init>(FileOutputStream.java:72)
                                                                  at com.zulip.android.activities.PhotoViewActivity$DownloadImage.doInBackground(PhotoViewActivity.java:174)
                                                                  at com.zulip.android.activities.PhotoViewActivity$DownloadImage.doInBackground(PhotoViewActivity.java:145)
                                                                  at android.os.AsyncTask$2.call(AsyncTask.java:295)
                                                                  at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                  at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
                                                                  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                  at java.lang.Thread.run(Thread.java:818)
                                                               Caused by: android.system.ErrnoException: open failed: EACCES (Permission denied)
                                                                  at libcore.io.Posix.open(Native Method)
                                                                  at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
                                                                  at libcore.io.IoBridge.open(IoBridge.java:438)
                                                                  at java.io.FileOutputStream.<init>(FileOutputStream.java:87) 
                                                                  at java.io.FileOutputStream.<init>(FileOutputStream.java:72) 
                                                                  at com.zulip.android.activities.PhotoViewActivity$DownloadImage.doInBackground(PhotoViewActivity.java:174) 
                                                                  at com.zulip.android.activities.PhotoViewActivity$DownloadImage.doInBackground(PhotoViewActivity.java:145) 
                                                                  at android.os.AsyncTask$2.call(AsyncTask.java:295) 
                                                                  at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                                                                  at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234) 
                                                                  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
                                                                  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
                                                                  at java.lang.Thread.run(Thread.java:818) 
```

File: app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+            case R.id.open:
+                Intent i = new Intent(Intent.ACTION_VIEW);
+                i.setData(Uri.parse(url));
+                startActivity(i);
+                break;
+            case R.id.copy_link:
+                copyLink(url);
+                break;
+            case R.id.download:
+                downloadImage(url);
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: is a `break;` missing here?

Comment (saketkumar95) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: @Sam1301 ofcourse It is. I still don't know how it successfully ran :O 
Thanks for pointing it out. I don't know how i missed it.

Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/PhotoViewActivity.java]: cool, no worries :)

**Comment by saketkumar95** [comment]
@kunall17 I guess everything is done from my side now. 
Review and let me for any changes. :)

**Comment by kunall17** [comment]
I tried this one, and after asking for permissions it doesn't start download i have to click the option again!

**Comment by saketkumar95** [comment]
@kunall17 Done!

**Comment by kunall17** [comment]
@timabbott LGTM!

**Comment by timabbott** [comment]
@saketkumar95 can you fix the commit messages for your two commits?

**Comment by saketkumar95** [comment]
@timabbott Done! :)

**Comment by Sam1301** [comment]
@saketkumar95 we recently added more checks to lint, maybe try rebasing branch with current master to resolve the lint errors.

**Comment by saketkumar95** [comment]
@kunall17 @Sam1301 Still failing. I don't know whats going wrong here.

**Comment by saketkumar95** [comment]
@timabbott 

- Removed the copies of commits.
- Lint issue resolved.

everything is done now. :)

**Comment by timabbott** [comment]
Merged, thanks @saketkumar95!!

**Pull request closed without merging**